apps/emqx_authn/src/emqx_authn_initializer.erl

```diff
@@ -0,0 +1,133 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2021 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%--------------------------------------------------------------------
+
+-module(emqx_authn_initializer).
+
+-behaviour(gen_statem).
+
+-define(DEFAULT_INTERVAL, 5).
+
+-include("emqx_authn.hrl").
+-include_lib("snabbkaffe/include/snabbkaffe.hrl").
+
+%% APIs
+-export([start_link/0,
+         start_link/1,
+         initialize/0,
+         deinitialize/0
+        ]).
+
+%% gen_statem callbacks
+-export([init/1,
+         callback_mode/0
+        ]).
+
+%% state functions
+-export([waiting_for_process/3,
+         initializing/3,
+         idle/3
+        ]).
+
+%%--------------------------------------------------------------------
+%% External interface
+%%--------------------------------------------------------------------
+
+start_link() ->
+        start_link(
+          #{name => ?AUTHN,
+            interval => ?DEFAULT_INTERVAL,
+            init_fun => fun(_) -> initialize() end}).
+
+start_link(#{} = Opts) ->
+        gen_statem:start_link(?MODULE, Opts, []).
+
+initialize() ->
+    _ = ?AUTHN:register_providers(emqx_authn:providers()),
+
+    lists:foreach(
+      fun({ChainName, RawAuthConfigs}) ->
+              AuthConfig = emqx_authn:check_configs(RawAuthConfigs),
+              ?AUTHN:initialize_authentication(
+                 ChainName,
+                 AuthConfig)
+      end,
+      chain_configs()).
+
+deinitialize() ->
+    ok = ?AUTHN:deregister_providers(provider_types()).
```

Comment:
Deinitialization and `emqx_authn` application stop removes providers, but not authenticators belonging to them.

Is this the intended behaviour?